lib/engine/chelsea_bridge.ex

```diff
@@ -2,56 +2,127 @@ defmodule Engine.ChelseaBridge do
 
+  alias Signs.Utilities.EtsUtils
+
+  # API constants
+  @base_api_url "https://www.chelseabridgesys.com/api/api/"
+  @api_status_endpoint "BridgeRealTime"
+  @api_token_endpoint "token"
+
+  @bridge_table_name :bridge_status
+
+  @type token :: %{
+          value: String.t() | nil,
+          expiration: DateTime.t() | nil
+        }
+
+  @type state :: %{
+          table: :ets.tab(),
+          token: token
+        }
+
   @callback bridge_status() :: %{raised: boolean() | nil, estimate: DateTime.t() | nil}
-  def bridge_status() do
-    case :ets.lookup(:bridge_status, :value) do
+  def bridge_status(ets_table_name \\ @bridge_table_name) do
+    case :ets.lookup(ets_table_name, :value) do
       [{:value, data}] -> data
       _ -> %{raised?: nil, estimate: nil}
     end
   end
 
-  def start_link(_) do
-    GenServer.start_link(__MODULE__, nil, name: __MODULE__)
+  def start_link(opts \\ []) do
+    name = opts[:gen_server_name] || __MODULE__
+    GenServer.start_link(__MODULE__, opts, name: name)
   end
 
-  def init(_) do
+  def init(opts) do
+    bridge_ets_table =
+      :ets.new(opts[:bridge_ets_table_name] || @bridge_table_name, [
+        :named_table,
+        read_concurrency: true
+      ])
+
+    state = %{
+      table: bridge_ets_table,
+      token: %{value: nil, expiration: nil}
+    }
+
     send(self(), :update)
-    :ets.new(:bridge_status, [:named_table, read_concurrency: true])
-    {:ok, %{}}
+
+    {:ok, state}
   end
 
   def handle_info(:update, state) do
     Process.send_after(self(), :update, 60_000)
-    url = Application.get_env(:realtime_signs, :chelsea_bridge_url)
-    auth = Application.get_env(:realtime_signs, :chelsea_bridge_auth)
     http_client = Application.get_env(:realtime_signs, :http_client)
+    now = Timex.now()
+
+    token =
+      if(state.token.expiration == nil or DateTime.compare(state.token.expiration, now) == :lt) do
+        update_api_token(now)
+      else
+        state.token
+      end
 
     with {:ok, %{status_code: 200, body: body}} <-
-           http_client.get(url, [{"Authorization", "Basic #{auth}"}]),
-         {:ok, data} <- Jason.decode(body),
-         %{"bridge" => %{"bridgeStatusId" => %{"status" => status}}} <- data do
-      estimate =
-        case data do
-          %{"lift_estimate" => %{"estimate_time" => estimate}} ->
-            String.replace(estimate, ~r/\.\d+$/, "")
-            |> Timex.parse!("{YYYY}-{M}-{D} {h24}:{m}:{s}")
-            |> DateTime.from_naive!("America/New_York")
-
-          _ ->
-            nil
-        end
-
-      :ets.insert(:bridge_status, {:value, %{raised?: status == "Raised", estimate: estimate}})
+           http_client.get("#{@base_api_url}#{@api_status_endpoint}", [
+             {"Authorization", "Bearer #{token.value}"}
+           ]),
+         {:ok, data} <- Jason.decode(body) do
+      EtsUtils.write_ets(
+        state.table,
+        %{
+          :value => %{
+            raised?: Map.get(data, "liftInProgress"),
```

Comment:
Although it's probably unlikely, these accesses will return `nil` if the data ever looks different from what we expect, which would probably cause a crash. We could add a little extra safety by translating this into a pattern match in the `with` clause above, which would instead let us fall through to the error logging below.

Side nit: In cases where we do just want to get a map value with no default, `data[key]` is equivalent.

Reply:
Great point! Updated to pattern match in the second step of the `with` clause as well to directly pull out `liftInProgress` and `estimatedDurationInMinutes`